extensions/amp-ad/0.1/amp-ad-ui.js

```diff
@@ -0,0 +1,159 @@
+   * ad server.
+   */
+  LOADED_NO_CONTENT: 3,
+};
+
+export class AmpAdUIHandler {
+
+  /**
+   * @param {!AMP.BaseElement} baseInstance
+   */
+  constructor(baseInstance) {
+    /** @private {!AMP.BaseElement} */
+    this.baseInstance_ = baseInstance;
+
+    /** {number} */
+    this.state = AdDisplayState.NOT_LAID_OUT;;
+  }
+
+  /**
+   * TODO(@zhouyx): Add ad tag to the ad.
+   */
+  init() {
+  }
+
+  /**
+   * Exposed function to ad that enable them to set UI to correct display state
+   * @param {number} state
+   */
+  setDisplayState(state) {
+    if (this.state == AdDisplayState.NOT_LAID_OUT) {
+      // Once unlayout UI applied, only another layout will change the UI again
+      if (state != AdDisplayState.LOADING) {
+        return;
+      }
+    }
+    switch (state) {
+      case AdDisplayState.LOADING:
+        this.displayLoadingUI_();
+        break;
+      case AdDisplayState.LOADED_RENDER_START:
+        this.displayRenderStartUI_();
+        break;
+      case AdDisplayState.LOADED_NO_CONTENT:
+        this.displayNoContentUI_();
+        break;
+      case AdDisplayState.NOT_LAID_OUT:
+        this.displayUnlayoutUI_();
+        break;
+      default:
+        dev().assert(state, 'state is not supported');
+    }
+  }
+
+  /**
+   * TODO(@zhouyx): apply placeholder, add ad loading indicator
+   * @private
+   */
+  displayLoadingUI_() {
+    this.state = AdDisplayState.LOADING;
+  }
+
+  /**
+   * TODO(@zhouyx): remove ad loading indicator
+   * @private
+   */
+  displayRenderStartUI_() {
+    this.state = AdDisplayState.LOADED_RENDER_START;
+  }
+
+  /**
+   * Apply UI for laid out ad with no-content
+   * If fallback exist try to display provided fallback
+   * Else try to collapse the ad (Note: may not succeed)
+   * TODO(@zhouyx): apply fallback, remove ad loading indicator
+   * @private
+   */
+  displayNoContentUI_() {
+    if (this.baseInstance_.getFallback()) {
+      this.baseInstance_.deferMutate(() => {
+        if (this.state == AdDisplayState.NOT_LAID_OUT) {
+          // If already unlaid out, do not replace current placeholder then.
+          return;
+        }
+        this.baseInstance_.togglePlaceholder(false);
+        this.baseInstance_.toggleFallback(true);
+        this.state = AdDisplayState.LOADED_NO_CONTENT;
+      });
+    } else {
+      this.baseInstance_.attemptChangeHeight(0).then(() => {
+        this.baseInstance_./*OK*/collapse();
+        this.state = AdDisplayState.LOADED_NO_CONTENT;
+      }, () => {
+        this.state = AdDisplayState.LOADED_NO_CONTENT;
+      });
+    }
+  }
+
+  /**
+   * Apply UI for unlaid out ad
+   * Hide fallback and show placeholder if exists
+   * Once unlayout UI applied, only another layout will change the UI again
+   * TODO(@zhouyx): remove ad loading indicator
+   * @private
+   */
+  displayUnlayoutUI_() {
+    this.state = AdDisplayState.NOT_LAID_OUT;
+    this.baseInstance_.deferMutate(() => {
+      this.baseInstance_.togglePlaceholder(true);
```

Comment:
Only another layout can change the state from `NOT_LAID_OUT` to any other state. I think it would be super rare cases since even we layout the ad again, it has low priority and can't finish before `deferMutate`. But sure it would be safe to check again.